Сreated calculate percentage function and tests for it

Yevheniia-Rudenko opened the pull request:
---
name: solution review 
about: A template PR for code review with a checklist
---

<!--
  make this PR easy to find:

  - assign: yourself
  - link it to the corresponding challenge issue on the project board
-->

<!-- describe the challenge you will solve -->

## Behavior

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [ ] There are tests for defensive assertions
- [ ] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations

## Strategy

### Do's

- [x] Variable names help to understand the strategy
- [ ] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [ ] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

abeddost reviewed (commented):
Hey Yeveheniia, I see this error: "Some checks were not successful." Do you have any idea why do we get these errors?

![image](https://github.com/user-attachments/assets/f22e903e-0a70-49e8-ac8c-cb801581977a)

abeddost commented:
Hey Yevheniia, nice job! I can’t merge the pull request because some checks didn’t pass.

abeddost reviewed (commented):
Hey Yeveheniia, great job! I've made some suggestions for you to review.

Yevheniia-Rudenko commented:
@abeddost  Thank you so much for your review! I will work on addressing the CI checks and incorporating your suggestions, as I really liked them! I’ll let you know as soon as I submit any changes.

abeddost commented:
You're welcome! Glad you found the suggestions helpful.

Yevheniia-Rudenko commented:
Hi @abeddost, I resolved all CI check issues! The main problem was with ruff formatting. After reinstalling ruff and running the ruff format command, everything worked.
The second issue was with function names. After the last review, I updated the function names using the UI but forgot to update them throughout the code. Once I caught and fixed all instances, I passed all checks successfully.

abeddost reviewed (commented):
Hey @Yevheniia-Rudenko,  I’ve gone through your implementation, and the **calculate_percentage** function is really well done. It works as expected and calculates the percentage correctly for any given numerator and denominator. The test cases you’ve added cover a lot of different scenarios, and they do a good job of validating the main functionality.

One thing I would suggest is adding a few more **edge and defensive test cases** to make sure the function handles all situations. Here’s a list of some cases that could be useful to include:

**1. Test for 100% calculation
2. Test for negative numerator
3. Test for negative denominator
4. Test for percentage greater than 100
5. Test for non-numeric inputs (like strings, lists, or dictionaries as inputs)
6. Test for very small values
7. Test for very large values
8. Test for precision handling**

These additional tests would help cover all bases and make sure the function works well in all cases.

Yevheniia-Rudenko commented:
Hi @abeddost! I just pushed some changes based on your review, could you please take a look? Thank you!

abeddost reviewed (approved):
Hi @Yevheniia-Rudenko, I’ve reviewed your changes, and everything looks perfect! Thank you for your hard work and dedication. Great job!

I am merging this branch into the main now.

Yevheniia-Rudenko commented:
Thank you so much @abeddost
